rocsolver@4.5.2 ^fmt@8.1.0:8.1.1 build fails: fmt/ranges.h: error: no matching member function for call to 'format'

### Steps to reproduce the issue

`rocsolver@4.5.2 ^fmt@8.1.1` builds fails using:
* `spack@develop` (3d29e0d64139e11b0551eae8a25dd9f1232515e1 from `Mon Mar 7 11:58:39 2022 -0700`)
* Ubuntu 20.04, X86_64
* GCC 9.3.0

It looks like there should be a conflict for `rocsolver@4.5.2 ^fmt@8.1.1`. Same error w/ `^fmt@8.1.0

Concrete spec: [rocsolver.spec.json.txt](https://github.com/spack/spack/files/8200915/rocsolver.spec.json.txt)

Concretization:
```
rocsolver@4.5.2%gcc@9.3.0~ipo+optimal amdgpu_target=gfx906:xnack- build_type=Release arch=linux-ubuntu20.04-x86_64
    ^cmake@3.22.2%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
        ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
            ^pkgconf@1.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^openssl@1.1.1m%gcc@9.3.0~docs certs=system arch=linux-ubuntu20.04-x86_64
            ^perl@5.34.0%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
                ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
                ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
                    ^diffutils@3.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libiconv@1.16%gcc@9.3.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64
    ^fmt@8.1.1%gcc@9.3.0~ipo+pic~shared build_type=RelWithDebInfo cxxstd=11 arch=linux-ubuntu20.04-x86_64
    ^hip@4.5.2%gcc@9.3.0~ipo build_type=Release patches=24ca25a,2a41904 arch=linux-ubuntu20.04-x86_64
        ^comgr@4.5.2%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
            ^llvm-amdgpu@4.5.2%gcc@9.3.0~ipo~link_llvm_dylib~llvm_dylib+openmp+rocm-device-libs build_type=Release patches=d999f3b arch=linux-ubuntu20.04-x86_64
                ^hwloc@2.7.0%gcc@9.3.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
                    ^libpciaccess@0.16%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^m4@1.4.19%gcc@9.3.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
                                ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^util-macros@1.19.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxml2@2.9.12%gcc@9.3.0~python arch=linux-ubuntu20.04-x86_64
                        ^xz@5.2.5%gcc@9.3.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^libelf@0.8.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^perl-data-dumper@2.173%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^python@3.8.12%gcc@9.3.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
                    ^expat@2.4.6%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-x86_64
                        ^libbsd@0.11.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libmd@1.0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
                        ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libffi@3.4.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^sqlite@3.37.2%gcc@9.3.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
                    ^util-linux-uuid@2.37.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^z3@4.8.14%gcc@9.3.0~gmp~ipo~python build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
            ^rocm-cmake@4.5.2%gcc@9.3.0~ipo+ldconfig build_type=Release arch=linux-ubuntu20.04-x86_64
        ^hsa-rocr-dev@4.5.2%gcc@9.3.0+image~ipo+shared build_type=Release patches=71e6851 arch=linux-ubuntu20.04-x86_64
            ^hsakmt-roct@4.5.2%gcc@9.3.0~ipo+shared build_type=Release patches=f926273 arch=linux-ubuntu20.04-x86_64
                ^libdrm@2.4.110%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^docbook-xml@4.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^docbook-xsl@1.79.2%gcc@9.3.0 patches=a92c397 arch=linux-ubuntu20.04-x86_64
                    ^libpthread-stubs@0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^meson@0.61.2%gcc@9.3.0 patches=aa6c50d arch=linux-ubuntu20.04-x86_64
                        ^ninja@1.10.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-pip@21.3.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-setuptools@59.4.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^py-wheel@0.37.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^numactl@2.0.14%gcc@9.3.0 patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu20.04-x86_64
                    ^autoconf@2.69%gcc@9.3.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
                    ^automake@1.16.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^xxd-standalone@8.2.1201%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^mesa@21.3.1%gcc@9.3.0+glx~llvm+opengl~opengles+osmesa~strip buildtype=debugoptimized default_library=shared patches=ada85be arch=linux-ubuntu20.04-x86_64
            ^bison@3.8.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^flex@2.6.3%gcc@9.3.0+lex~nls arch=linux-ubuntu20.04-x86_64
                ^findutils@4.8.0%gcc@9.3.0 patches=440b954 arch=linux-ubuntu20.04-x86_64
            ^glproto@1.4.17%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libunwind@1.5.0%gcc@9.3.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib components=none libs=shared,static arch=linux-ubuntu20.04-x86_64
            ^libx11@1.7.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^inputproto@2.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^kbproto@1.0.7%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libxcb@1.14%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxau@1.0.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^xproto@7.0.31%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxdmcp@1.1.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^xcb-proto@1.14.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^xextproto@7.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^xtrans@1.3.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libxext@1.3.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libxt@1.1.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libice@1.0.9%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libsm@1.2.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-mako@1.1.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^py-markupsafe@2.0.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^xrandr@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libxrandr@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxrender@0.9.10%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^renderproto@0.11.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^randrproto@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^rocminfo@4.5.2%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
        ^roctracer-dev-api@4.5.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
    ^rocblas@4.5.2%gcc@9.3.0~ipo build_type=Release patches=6a52b28 tensile_architecture=all arch=linux-ubuntu20.04-x86_64
        ^msgpack-c@3.1.1%gcc@9.3.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
        ^perl-file-which@1.22%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^py-msgpack@1.0.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^py-pyyaml@6.0%gcc@9.3.0+libyaml arch=linux-ubuntu20.04-x86_64
            ^libyaml@0.2.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-cython@0.29.24%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^py-virtualenv@20.10.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-backports-entry-points-selectable@1.1.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^py-setuptools-scm@6.3.2%gcc@9.3.0+toml arch=linux-ubuntu20.04-x86_64
                    ^py-packaging@21.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-pyparsing@3.0.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^py-tomli@1.2.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-distlib@0.3.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-filelock@3.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-platformdirs@2.4.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-six@1.16.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^rocm-smi-lib@4.5.2%gcc@9.3.0~ipo+shared build_type=Release patches=8bc40cc arch=linux-ubuntu20.04-x86_64
```

Reproducer (here using Docker image `ecpe4s/ubuntu20.04-runner-x86_64:2022-02-01`)
```
$> spack mirror add E4S https://cache.e4s.io
$> spack buildcache keys -it
$> spack install --cache-only --include-build-deps --only dependencies -f ./rocsolver.spec.json

$> spack install --no-cache --only package -f ./rocsolver.spec.json
...
==> Installing rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c
==> Fetching https://mirror.spack.io/_source-cache/archive/46/4639322bd1e77fedfdeb9032633bde6211a0b1cc16a612db7754f873f18a492f.tar.gz
==> No patches needed for rocsolver
==> rocsolver: Executing phase: 'cmake'
==> rocsolver: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'
...
     4814    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/auxiliary/rocauxiliary_orgtr_ungtr.cpp:5:
     4815    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/auxiliary/rocauxiliary_orgtr_ungtr.hpp:12:
     4816    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/auxiliary/rocauxiliary_orgql_ungql.hpp:12:
     4817    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/auxiliary/rocauxiliary_larfb.hpp:12:
     4818    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/include/rocblas.hpp:15:
     4819    In file included from /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/include/rocsolver_logger.hpp:9:
  >> 4820    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/fmt-8.1.1-mswoitr42xb2xrytqwp2jgl535d22dcg/include/fmt/ranges.h:737:21: error: no matching member function for call to 'format'
     4821                       .format(std::get<sizeof...(T) - N>(value.tuple), ctx);
     4822                       ~^~~~~~
     4823    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/fmt-8.1.1-mswoitr42xb2xrytqwp2jgl535d22dcg/include/fmt/ranges.h:741:14: note: in instantiation of function template specialization
              'fmt::formatter<fmt::tuple_join_view<char, rocsolver_logvalue<const char *> &, rocsolver_logvalue<rocblas_fill_> &, rocsolver_logvalue<const char *> &, rocsolver_logvalue<int> &, ro
             csolver_logvalue<const char *> &, rocsolver_logvalue<int> &>>::do_format<fmt::basic_format_context<fmt::appender, char>, 1>' requested here
     4824          return do_format(value, ctx, std::integral_constant<size_t, N - 1>());
     4825                 ^
     4826    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/fmt-8.1.1-mswoitr42xb2xrytqwp2jgl535d22dcg/include/fmt/ranges.h:741:14: note: in instantiation of function template specialization
              'fmt::formatter<fmt::tuple_join_view<char, rocsolver_logvalue<const char *> &, rocsolver_logvalue<rocblas_fill_> &, rocsolver_logvalue<const char *> &, rocsolver_logvalue<int> &, ro
             csolver_logvalue<const char *> &, rocsolver_logvalue<int> &>>::do_format<fmt::basic_format_context<fmt::appender, char>, 2>' requested here

     ...

     4849        return rocsolver_orgtr_ungtr_impl<float>(handle, uplo, n, A, lda, ipiv);
     4850               ^
     4851    /tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-src/library/src/include/rocsolver_logvalue.hpp:40:10: note: candidate function template not v
             iable: 'this' argument has type 'const __tuple_element_t<5UL, tuple<fmt::formatter<rocsolver_logvalue<const char *>>, fmt::formatter<rocsolver_logvalue<rocblas_fill>>, fmt::formatter
             <rocsolver_logvalue<const char *>>, fmt::formatter<rocsolver_logvalue<int>>, fmt::formatter<rocsolver_logvalue<const char *>>, fmt::formatter<rocsolver_logvalue<int>>>>' (aka 'const
             fmt::formatter<rocsolver_logvalue<int>>'), but method is not marked const
     4852        auto format(rocsolver_logvalue<T> wrapper, FormatCtx& ctx)
     4853             ^
     4854    6 errors generated when compiling for gfx906.
  >> 4855    make[2]: *** [library/src/CMakeFiles/rocsolver.dir/build.make:303: library/src/CMakeFiles/rocsolver.dir/auxiliary/rocauxiliary_orgtr_ungtr.cpp.o] Error 1
     4856    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-build-bglnzqu'
  >> 4857    make[1]: *** [CMakeFiles/Makefile2:134: library/src/CMakeFiles/rocsolver.dir/all] Error 2
     4858    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-rocsolver-4.5.2-bglnzqutgp7hj52chfekrwu3i5ivpp6c/spack-build-bglnzqu'
  >> 4859    make: *** [Makefile:159: all] Error 2
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.17.1-1474-3d29e0d641
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8200931/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/8200932/spack-build-out.txt)

@arjun-raj-kuppala @haampie @srekolam 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/rocsolver/package.py
+++ b/var/spack/repos/builtin/packages/rocsolver/package.py
@@ -43,7 +43,7 @@ class Rocsolver(CMakePackage):
 
     depends_on('cmake@3.8:', type='build', when='@4.1.0:')
     depends_on('cmake@3.5:', type='build')
-    depends_on('fmt@7.1.3:', type='build', when='@4.5.0:')
+    depends_on('fmt@7:8.0.1', type='build', when='@4.5.0:')
 
     depends_on('googletest@1.10.0:', type='test')
     depends_on('netlib-lapack@3.7.1:', type='test')

